File: examples/android-compose/src/main/kotlin/xyz/superfunction/spfn/example/ExampleReceipt.kt
// SPFN Mobile — what one cell run left behind.
//
// The shape is tools/harness/.../HarnessReceipt.kt's and the reasons are the same: the
// JSON is assembled by hand so that every field which reaches the file is named in this
// class, and the timestamp is formatted with an explicit UTC zone and Locale.ROOT because
// a default locale can carry a non-Gregorian calendar or non-ASCII digits (P9), and with
// SimpleDateFormat rather than java.time because java.time arrives on Android at API 26
// and this app's floor is 24 (P14).
//
// **Nothing here is a credential.** A cell id, a fixture name, a stack depth, a state
// word, and the two versions this build is. There is nothing else to carry: this app
// never enrols and never holds a key.
//
// It is a copy of the harness's pattern and not a shared helper. The two apps write
// different receipts about different things into different directories, and coupling them
// would make one app's evidence format a constraint on the other's.

package xyz.superfunction.spfn.example

import android.content.Context
import java.io.File
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.TimeZone

class ExampleReceipt(
    /** The cell the launch named, or `none` when it named nothing this app knows. */
    val cell: String,
    /** The seeding that cell ran under, or `none`. */
    val fixture: String,
    /** How deep the flow's stack was when the receipt was written. */
    val stackDepth: Int,
    val timestampMillis: Long,
    val sdkVersion: String,
    val contractVersion: String
)
{
    /**
     * `receipt-<cell>-<epochMillis>.json`.
     *
     * Milliseconds, not seconds, for the reason the harness records: two runs of one cell
     * that finished inside a second shared a name, and the second destroyed the first's
     * evidence — a failure mode where the more you run, the less you have.
     */
    fun fileName(): String = "receipt-$cell-$timestampMillis.json";

    /** The receipt as JSON, in a fixed field order this class is the whole list of. */
    fun toJson(): String = buildString {
        append("{\n");
        appendField("schema", SCHEMA, last = false);
        appendField("platform", PLATFORM, last = false);
        appendField("cell", cell, last = false);
        appendField("fixture", fixture, last = false);
        append("  \"stackDepth\": ").append(stackDepth).append(",\n");
        appendField("timestamp", timestamp(timestampMillis), last = false);
        appendField("sdkVersion", sdkVersion, last = false);
        appendField("contractVersion", contractVersion, last = true);
        append("}\n");
    };

    private fun StringBuilder.appendField(name: String, value: String, last: Boolean)
    {
        append("  \"").append(name).append("\": ");
        appendJsonString(value);
        append(if (last) "\n" else ",\n");
    }

    companion object
    {
        const val SCHEMA: String = "spfn-ui-cell-receipt/1";
        const val PLATFORM: String = "android";

        /** Hex digits as data, so no formatter and therefore no locale is involved (P9). */
        private const val HEX: String = "0123456789abcdef";

        /** The run's instant, ISO-8601 in UTC. See this file's header for both reasons. */
        fun timestamp(millis: Long): String
        {
            val format = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'", Locale.ROOT);
            format.timeZone = TimeZone.getTimeZone("UTC");
            return format.format(Date(millis));
        }

        /** One JSON string literal, escaped to printable ASCII. */
        fun StringBuilder.appendJsonString(value: String)
        {
            append('"');
            for (character in value)
            {
                when
                {
                    character == '"' -> append("\\\"")
                    character == '\\' -> append("\\\\")
                    character.code in 0x20..0x7e -> append(character)
                    else -> appendUnicodeEscape(character)
                }
            }
            append('"');
        }

        private fun StringBuilder.appendUnicodeEscape(character: Char)
        {
            append("\\u");
            append(HEX[(character.code shr 12) and 0xf]);
            append(HEX[(character.code shr 8) and 0xf]);
            append(HEX[(character.code shr 4) and 0xf]);
            append(HEX[character.code and 0xf]);
        }
    }
}

/**
 * Where receipts land, and what happens when they cannot.
 *
 * A `receipts` subdirectory of the external files directory, which is the one an `adb pull`
 * reaches without root. Its own subdirectory, and a different app id from the harness's, so
 * neither run can be mistaken for the other's evidence.
 */
class ExampleReceiptStore(private val context: Context)
{
    /**
     * Writes one receipt and answers its file name.
     *
     * A failure to write is raised, never swallowed: a run whose receipt silently did not
     * appear is indistinguishable from a run that never happened (P7).
     */
    fun write(receipt: ExampleReceipt): String
    {
        val root = context.getExternalFilesDir(null)
            ?: throw IllegalStateException("no external files directory; nowhere to write a receipt");
        val directory = File(root, "receipts");
        if (!directory.exists() && !directory.mkdirs())
        {
            throw IllegalStateException("the receipts directory could not be created");
        }
        val file = File(directory, receipt.fileName());
        file.writeText(receipt.toJson());
        return file.name;
    }
}

File: examples/ios-swiftui/Sources/ExampleReceipt.swift
// SPFN Mobile — what one cell run left behind.
//
// Counterpart of
// examples/android-compose/src/main/kotlin/xyz/superfunction/spfn/example/ExampleReceipt.kt,
// and, like it, a copy of tools/harness's receipt pattern rather than a shared helper: the
// two apps write different receipts about different things into different directories, and
// coupling them would make one app's evidence format a constraint on the other's.
//
// The JSON is assembled by hand so that every field which reaches the file is named in
// this type. **Nothing here is a credential.** A cell id, a fixture name, a stack depth
// and the two versions this build is — this app never enrols and never holds a key.

import Foundation

struct ExampleReceipt: Sendable
{
    static let schema = "spfn-ui-cell-receipt/1"
    static let platform = "ios"

    /// The cell the launch named, or `none` when it named nothing this app knows.
    let cell: String

    /// The seeding that cell ran under, or `none`.
    let fixture: String

    /// How deep the flow's stack was when the receipt was written.
    let stackDepth: Int

    let timestampMillis: Int64
    let sdkVersion: String
    let contractVersion: String

    /// `receipt-<cell>-<epochMillis>.json`.
    ///
    /// Milliseconds, not seconds, for the reason the harness records: two runs of one cell
    /// that finished inside a second shared a name, and the second destroyed the first's
    /// evidence — a failure mode where the more you run, the less you have.
    var fileName: String
    {
        "receipt-\(cell)-\(timestampMillis).json"
    }

    /// The receipt as JSON, in a fixed field order this type is the whole list of.
    func toJSON() -> String
    {
        var out = "{\n"
        out += field("schema", Self.schema, last: false)
        out += field("platform", Self.platform, last: false)
        out += field("cell", cell, last: false)
        out += field("fixture", fixture, last: false)
        out += "  \"stackDepth\": \(stackDepth),\n"
        out += field("timestamp", Self.timestamp(timestampMillis), last: false)
        out += field("sdkVersion", sdkVersion, last: false)
        out += field("contractVersion", contractVersion, last: true)
        out += "}\n"
        return out
    }

    private func field(_ name: String, _ value: String, last: Bool) -> String
    {
        "  \"\(name)\": " + Self.jsonString(value) + (last ? "\n" : ",\n")
    }

    /// The run's instant, ISO-8601 in UTC.
    ///
    /// An explicit UTC zone and `Locale(identifier: "en_US_POSIX")`, both load-bearing: a
    /// default locale can carry a non-Gregorian calendar and can render digits in a
    /// non-ASCII script, and neither is a receipt anything can read
    /// (docs/IMPLEMENTATION-PITFALLS.md P9).
    static func timestamp(_ millis: Int64) -> String
    {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ss'Z'"
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.timeZone = TimeZone(identifier: "UTC")
        return formatter.string(from: Date(timeIntervalSince1970: Double(millis) / 1000))
    }

    /// One JSON string literal, escaped to printable ASCII.
    static func jsonString(_ value: String) -> String
    {
        var out = "\""
        for scalar in value.unicodeScalars
        {
            switch scalar
            {
            case "\"":
                out += "\\\""
            case "\\":
                out += "\\\\"
            case let printable where printable.value >= 0x20 && printable.value <= 0x7e:
                out.unicodeScalars.append(printable)
            default:
                out += unicodeEscape(scalar.value)
            }
        }
        return out + "\""
    }

    /// Hex digits as data, so no formatter and therefore no locale is involved (P9).
    private static func unicodeEscape(_ value: UInt32) -> String
    {
        let digits = Array("0123456789abcdef")
        var out = "\\u"
        for shift in stride(from: 12, through: 0, by: -4)
        {
            out.append(digits[Int((value >> UInt32(shift)) & 0xf)])
        }
        return out
    }
}

/// Where receipts land, and what happens when they cannot.
///
/// A `receipts` subdirectory of the app's Documents directory, which is what
/// `UIFileSharingEnabled` exposes to Finder and to the Files app. Its own subdirectory,
/// and a different bundle id from the harness's, so neither run can be mistaken for the
/// other's evidence.
struct ExampleReceiptStore: Sendable
{
    /// Writes one receipt and answers its file name.
    ///
    /// A failure to write is thrown, never swallowed: a run whose receipt silently did not
    /// appear is indistinguishable from a run that never happened (P7).
    func write(_ receipt: ExampleReceipt) throws -> String
    {
        let documents = try FileManager.default.url(
            for: .documentDirectory,
            in: .userDomainMask,
            appropriateFor: nil,
            create: true
        )
        let directory = documents.appendingPathComponent("receipts", isDirectory: true)
        try FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
        let file = directory.appendingPathComponent(receipt.fileName)
        try receipt.toJSON().write(to: file, atomically: true, encoding: .utf8)
        return receipt.fileName
    }
}
